### ISSUE TITLE
Does not work since Pidgin 2.14.9

### ISSUE THREAD

**Issue by SammysHP** [kgraefe/pidgin-znc-helper#19]
Since "unknown date somewhere this year probably" timestamps from the playback buffer are not fixed anymore and also not stripped from the original message.

```
(16:52:31) ***: Buffer Playback...
(16:52:31) xxxxxx: xxxxxxxxxxxxxxxxxxxxx  [2022-05-21 16:28:14]
(16:52:31) ***: Playback Complete.
```

Pidgin 2.14.9
pidgin-znc-helper 1.9

**Comment by kgraefe** [comment]
What ZNC version is running on the server?

**Comment by kgraefe** [comment]
Sorry, that shouldn't matter. Since version 1.8 the plugin does not parse the message anymore but uses the `znc.in/server-time-iso` capability which has been around since ZNC 1.2. It got renamed to `server-time` in ZNC 1.7.0 but kept the old name for compatibility. The old name will be removed eventually but at the time is still around in the master branch of ZNC. So this shouldn't be the issue here.

Can you open the debug log (in the Pidgin Buddy List -> Help -> Debug Window) and reconnect the account?

Also can you please check that `Uses ZNC bouncer` is enabled in the account settings? Is `Authenticate with SASL` disabled?

**Comment by SammysHP** [comment]
znc 1.7.2-3 from Debian Buster which hasn't changed since 2019.

`Uses ZNC bouncer` is enabled for that account in Pidgin and SASL is only used between ZNC and the IRC networks as far as I remember. Will check the log when I'm back home.

**Comment by SammysHP** [comment]
Here's the debug log:

```
(18:29:21) account: Connecting to account liberachat@***.*****.***.
(18:29:21) connection: Connecting. gc = 0x562a862c0850
(18:29:21) dnsquery: Performing DNS lookup for ***.*****.***
(18:29:21) jabber: jabber_actions: have pep: YES
(18:29:21) dns: Wait for DNS child 1499 failed: Keine Kind-Prozesse
(18:29:21) dns: Wait for DNS child 1497 failed: Keine Kind-Prozesse
(18:29:21) dns: Wait for DNS child 1500 failed: Keine Kind-Prozesse
(18:29:21) dns: Wait for DNS child 1498 failed: Keine Kind-Prozesse
(18:29:21) dns: Created new DNS child 2200, there are now 1 children.
(18:29:21) dns: Successfully sent DNS request to child 2200
(18:29:21) dns: Got response for '***.*****.***'
(18:29:21) dnsquery: IP resolved for ***.*****.***
(18:29:21) proxy: Attempting connection to *:*:*:*:*:*:*
(18:29:21) proxy: Connecting to ***.*****.***:*** with no proxy
(18:29:21) proxy: Connection in progress
(18:29:21) proxy: Connecting to ***.*****.***:***.
(18:29:21) proxy: Connected to ***.*****.***:***.
(18:29:22) nss: SSL version 3.3 using 256-bit CHACHA20POLY1305 with 128-bit AEAD MAC
Server Auth: 2048-bit RSA, Key Exchange: 255-bit ECDHE, Compression: NULL
Cipher Suite Name: TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
(18:29:22) nss: subject=E=znc@host.unknown,CN=host.unknown,OU=znc issuer=E=znc@host.unknown,CN=host.unknown,OU=znc
(18:29:22) certificate/x509/tls_cached: Starting verify for ***.*****.***
(18:29:22) certificate/x509/tls_cached: Checking for cached cert...
(18:29:22) certificate/x509/tls_cached: ...Found cached cert
(18:29:22) nss/x509: Loading certificate from /home/sammyshp/.purple/certificates/x509/tls_peers/***.*****.***
(18:29:22) certificate/x509/tls_cached: Peer cert matched cached
(18:29:22) nss/x509: Exporting certificate to /home/sammyshp/.purple/certificates/x509/tls_peers/***.*****.***
(18:29:22) util: Writing file /home/sammyshp/.purple/certificates/x509/tls_peers/***.*****.***
(18:29:22) nss: Trusting E=znc@host.unknown,CN=host.unknown,OU=znc
(18:29:22) certificate: Successfully verified certificate for ***.*****.***
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 001 liberachat :Welcome to the Libera.Chat Internet Relay Chat Network SammysHP
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 002 liberachat :Your host is tantalum.libera.chat[93.158.237.2/6697], running version solanum-1.0-dev
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 003 liberachat :This server was created Mon Apr 18 2022 at 05:57:25 UTC
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 004 liberachat tantalum.libera.chat solanum-1.0-dev DGMQRSZaghilopsuwz CFILMPQSTbcefgijklmnopqrstuvz bkloveqjfI
(18:29:22) jabber: jabber_actions: have pep: YES
(18:29:22) connection: Activating keepalive.
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 252 liberachat 41 :IRC Operators online
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 254 liberachat 22536 :channels formed
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 265 liberachat 2528 2967 :Current local users 2528, max 2967
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 266 liberachat 45251 48018 :Current global users 45251, max 48018
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 250 liberachat :Highest connection count: 2968 (2967 clients) (114624 connections received)
(18:29:22) irc: Unrecognized message: :irc.znc.in 306 SammysHP :You have been marked as being away
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) g_log: purple_strreplace: assertion 'string != NULL' failed
(18:29:22) irc: Unrecognized message: :tantalum.libera.chat 305 SammysHP :You are no longer marked as being away
(18:29:26) util: Writing file accounts.xml to directory /home/sammyshp/.purple
(18:29:27) util: Writing file /home/sammyshp/.purple/accounts.xml
(18:29:27) util: Writing file status.xml to directory /home/sammyshp/.purple
(18:29:27) util: Writing file /home/sammyshp/.purple/status.xml
```

**Comment by kgraefe** [comment]
Great, there is no debug output in the plugin at all... :-\ I will have to prepare an intermediate version for you. Are you on Windows?

You're set up is exactly to what I am using on a daily basis, except for the ZNC version.

**Comment by SammysHP** [comment]
I'm on Arch linux, maintainer of the AUR package: https://aur.archlinux.org/packages/pidgin-znchelper

I might try downgrading pidgin and check the ZNC config. Not sure when exactly it started, after a server reboot or pidgin update.

**Comment by SammysHP** [comment]
Just inserted some debug messages and it looks like `irc_sending_text_cb()`, which registers the znc connection, is never called with `connection_state == PURPLE_CONNECTING`, it is always `PURPLE_CONNECTED`. Thus the connection handler will never be created.

edit: Downgraded to pidgin and libpurple 2.14.8 and it is working again. In the logs I see a call of `irc_sending_text_cb()` with `PASS :…` which is missing in 2.14.9.

**Comment by SammysHP** [comment]
This seems to be the cause of the issue: https://keep.imfreedom.org/pidgin/pidgin/rev/c49dcf00bee6

`irc_send()` was replaced by `irc_priority_send()` or `do_send()`, but the `irc-sending-text` event is emitted only by `irc_send()` which is not involved in the chain anymore.

**Comment by kgraefe** [comment]
uh, okay. Thank you for your analysis. I'll look into it as soon as possible. I hope I can find another way to hook into the IRC authentication protocol (which is where capabilities are negotiated).

Also I lied to you: since I switched to Ubuntu I don't use latest Pidgin anymore. Sorry for that.

**Comment by SammysHP** [comment]
I wonder if this might be a regression that should be reported upstream. I couldn't find any clear documentation of `irc-sending-text`, but the behavior shouldn't change, especially not in a patch version. Do you have ways to report this upstream? I don't have an account at their bug tracker yet.

**Comment by kgraefe** [comment]
@grimmy can you help?

**Comment by kgraefe** [comment]
I created an upstream bug: https://issues.imfreedom.org/issue/PIDGIN-17626

**Comment by grimmy** [comment]
Sorry about that. We have an review request up (see the upstream bug) that should fix the issue in the next version which will probably be pretty soon.

**Comment by kgraefe** [comment]
Thank you. I still had to change the plugin a bit as the capability negotiation was rate limited when sent through `send_raw()`: d240b634b59b12fef8774fb07662e6e027a5620c

**Comment by SammysHP** [comment]
Thanks for your work. I will try it with the next release of Pidgin.

(Maybe Pidgin needs another option to turn off rate limiting. My ZNC does this for me, now lines are throttled twice.)

**Comment by grimmy** [comment]
> Thank you. I still had to change the plugin a bit as the capability negotiation was rate limited when sent through `send_raw()`: [d240b63](https://github.com/kgraefe/pidgin-znc-helper/commit/d240b634b59b12fef8774fb07662e6e027a5620c)

Crap that's my fault I think. When I wrote the rate limiter I clearly didn't think anyone else needed to sending messages during the registration time and iirc I don't even start the rate limiter until after that which is why it's so late..

That said, what you're doing should be fine.

**Comment by kgraefe** [comment]
> Maybe Pidgin needs another option to turn off rate limiting. My ZNC does this for me, now lines are throttled twice.

There are new settings in the advanced account preferences you could use to effectively switch it off (`Seconds between sending messages` and `Maximum messages to send at once`)

**Comment by kgraefe** [comment]
Fix released in [version 1.10](https://github.com/kgraefe/pidgin-znc-helper/releases/tag/v1.10) and [Pidgin 2.14.10](https://pidgin.im/posts/2022-06-2.14.10-released/)
